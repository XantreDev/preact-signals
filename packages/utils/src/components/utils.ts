import { ReadonlySignal, Signal, useComputed, useSignal } from "@preact-signals/unified-signals";
import { useRef } from "react";

const isSignal = <T>(value: T | ReadonlySignal<T>): value is Signal<T> =>
  value instanceof Signal;

export const useLiveSignal = <T>(maybeSignal: T | ReadonlySignal<T>) => {
  const signalValue = useSignal(maybeSignal);
  if (signalValue.peek() !== maybeSignal) {
    signalValue.value = maybeSignal;
  }
  return useComputed(() => {
    const value = signalValue.value;
    return isSignal(value) ? value.value : (value as T);
  });
};

const empty = Symbol("empty");
export const useConstant = <T>(fn: () => T) => {
  const ref = useRef<T | typeof empty>(empty);
  if (ref.current === empty) {
    ref.current = fn();
  }
  return ref.current;
};
