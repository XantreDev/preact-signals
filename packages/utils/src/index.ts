export * from "./$";
export {
  createFlatStore,
  flatStore,
  type FlatStore,
  type ReadonlyFlatStore
} from "./flat-store";
export * from "./resource";
export {
  accessorOfReactive,
  accessorOfSignal,
  reaction,
  setterOfSignal,
  stableAccessorOfSignal,
  toggleSignal,
  untracked,
  unwrapReactive,
  type Accessor,
  type AnyReactive,
  type CreateFunction,
  type Reactive,
  type Setter
} from "./utils";

declare const signalSymbol: unique symbol;
// @ts-expect-error
declare module "@preact/signals-core" {
  interface Signal<T> {
    /** flag for easier identification */
    [signalSymbol]: true;
  }
}
