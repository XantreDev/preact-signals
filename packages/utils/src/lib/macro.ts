import { type ReactiveRef } from "./$";

console.log(
  "To use macro, you need to use babel plugin. Check out the README for more info."
);

export type $$Type = <T>(value: T) => ReactiveRef<T>;

const createMacroError =
  (macroName: string): ((...args: any[]) => never) =>
  () => {
    throw new Error(
      `'${macroName}' function is only available at compile time`
    );
  };

/**
 * Macro function that is compile time shorthand for `$(() => value)`
 *
 * @example
 * ```tsx
 * import { $$ } from "@preact-signals/utils/macro";
 *
 * // This is equivalent to:
 * // const a = $(() => 1)
 * const a = $$(1)
 * ```
 *
 * @example A more complex example
 * ```tsx
 * import { $$ } from "@preact-signals/utils/macro";
 *
 * const a = signal(1)
 * const sig = signal(1)
 * const b = $$(a.value + sig.value)
 *
 * effect(() => {
 *   console.log(b.value)
 * })
 *
 * // logs 3
 * sig.value = 2
 *
 * // logs 4
 * a.value = 2
 * ```
 */
export const $$: $$Type = createMacroError("$$");

export type $StateMacroType = <T>(value: T) => T;

/**
 * Macro hook-function that allows you to create global reactive binding. Compile time wrapper around `deepSignal`
 *
 * @example
 * ```tsx
 * import { $state } from "@preact-signals/utils/macro";
 *
 * let a = $state(1)
 *
 * effect(() => {
 *  console.log(a)
 * })
 *
 * // logs 1
 * a = 2
 * ```
 *
 */
export const $state: $StateMacroType = createMacroError("$state");

/**
 * Macro hook-function that allows you to create deep reactive binding inside of component. Compile time wrapper around `useDeepSignal`
 *
 * @example
 * ```tsx
 * import { $useState } from "@preact-signals/utils/macro";
 *
 * const Counter = () => {
 *   let count = $useState(1)
 *
 *   return (
 *    <div>
 *     <button onClick={() => count++}>Increment</button>
 *     <p>{count}</p>
 *   </div>
 *  )
 * }
 * ```
 *
 */
export const $useState: $StateMacroType = createMacroError("$useState");

/**
 * Macro hook-function that creates reactive binding that is linked to a state that passed as argument. Compile time wrapper around `useSignalOfState`
 *
 * @example
 * ```tsx
 * import { $useLinkedState } from "@preact-signals/utils/macro";
 *
 * const Counter = (props: {count: number; onIncrement(): void}) => {
 *  // only allowed with const
 *  const count = $useLinkedState(props.count)
 *  // allows to use value in effect
 *  useSignalEffect(() => {
 *   console.log(count)
 *  })
 *
 *  return (
 *   <div>
 *    <button onClick={props.onIncrement}>Increment</button>
 *    <p>{props.count}</p>
 *   </div>
 *  )
 * }
 * ```
 */
export const $useLinkedState: $StateMacroType =
  createMacroError("$useLinkedState");

/**
 * Macro function that creates a derived reactive binding. Compile time wrapper around `computed`
 * @example
 * ```ts
 * import { $derived, $state } from "@preact-signals/utils/macro";
 *
 * let a = $state(1)
 * let b = $state(2)
 *
 * const c = $derived(a + b)
 * const doubleC = $derived(c * 2)
 *
 * console.log(c) // 3
 * console.log(doubleC) // 6
 * a = 3
 * console.log(c) // 5
 * console.log(doubleC) // 10
 * ```
 */
export const $derived: $StateMacroType = createMacroError("$derived");

/**
 * Macro function that creates a derived reactive binding. Compile time wrapper around `computed`
 *
 * @example
 * ```ts
 * import { $useDerived, $useState } from "@preact-signals/utils/macro";
 *
 * const Component = () => {
 *  let counter = $useState(0)
 *  const doubledCounter = $useDerived(counter * 2)
 *  const increment = () => counter++
 *
 *  return (
 *   <div>
 *    <button onClick={onIncrement}>Increment</button>
 *    <p>Counter {count}</p>
 *    <p>Doubled Counter {doubledCounter}</p>
 *   </div>
 *  );
 * }
 * ```
 */
export const $useDerived: $StateMacroType = createMacroError("$useDerived");
