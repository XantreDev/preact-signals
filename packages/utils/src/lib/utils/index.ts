export { untracked } from "@preact-signals/unified-signals";
export { isExplicitFalsy } from "./explicitFalsy";
export type * from "./explicitFalsy";
export { accessorOfSignal, stableAccessorOfSignal } from "./getter";
export { accessorOfReactive, unwrapReactive } from "./reactive";
export { setterOfSignal, toggleSignal } from "./setter";
export { reducerSignal, type Reducer, ReducerSignal } from "./reducer";
export { toSolidLikeSignal } from "./toSolidApi";
export type * from "./toSolidApi";
export type * from "./type";
export { writableRefOfArrayProp, writableRefOfObjectProp } from "./object";
export { reaction, rafReaction } from "./reaction";
export type * from "./reaction";
