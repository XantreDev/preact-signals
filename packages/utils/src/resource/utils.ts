import { NO_INIT } from "./constants";

/*@__NO_SIDE_EFFECTS__*/
export const isPromise = <T>(value: unknown): value is Promise<T> =>
  !!value && typeof value === "object" && value instanceof Promise;

/*@__NO_SIDE_EFFECTS__*/
export const removeNoInit = <T>(value: T | typeof NO_INIT) =>
  value !== NO_INIT ? value : undefined;
