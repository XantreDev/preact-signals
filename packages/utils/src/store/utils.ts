/**
 * Always return false.
 */
export const NO = () => false;

const onRE = /^on[^a-z]/;
export const isOn = (key: string) => onRE.test(key);

export const isModelListener = (key: string) => key.startsWith("onUpdate:");

export const extend = Object.assign;

export const remove = <T>(arr: T[], el: T) => {
  const i = arr.indexOf(el);
  if (i > -1) {
    arr.splice(i, 1);
  }
};

export const hasChanged = (value: any, oldValue: any): boolean =>
  !Object.is(value, oldValue);

const hasOwnProperty = Object.prototype.hasOwnProperty;
export const hasOwn = (
  val: object,
  key: string | symbol
): key is keyof typeof val => hasOwnProperty.call(val, key);

export const isArray = Array.isArray;
export const isMap = (val: unknown): val is Map<any, any> =>
  toTypeString(val) === "[object Map]";
export const isSet = (val: unknown): val is Set<any> =>
  toTypeString(val) === "[object Set]";

export const isDate = (val: unknown): val is Date =>
  toTypeString(val) === "[object Date]";
export const isRegExp = (val: unknown): val is RegExp =>
  toTypeString(val) === "[object RegExp]";
export const isFunction = (val: unknown): val is Function =>
  typeof val === "function";
export const isString = (val: unknown): val is string =>
  typeof val === "string";
export const isSymbol = (val: unknown): val is symbol =>
  typeof val === "symbol";
export const isObject = (val: unknown): val is Record<any, any> =>
  val !== null && typeof val === "object";

export const isPromise = <T = any>(val: unknown): val is Promise<T> => {
  return (
    (isObject(val) || isFunction(val)) &&
    isFunction((val as any).then) &&
    isFunction((val as any).catch)
  );
};
/**
 * Make a map and return a function for checking if a key
 * is in that map.
 * IMPORTANT: all calls of this function must be prefixed with
 * \/\*#\_\_PURE\_\_\*\/
 * So that rollup can tree-shake them if necessary.
 */
export function makeMap(
  str: string,
  expectsLowerCase?: boolean
): (key: string) => boolean {
  const map: Record<string, boolean> = Object.create(null);
  const list: Array<string> = str.split(",");
  for (let i = 0; i < list.length; i++) {
    map[list[i]!] = true;
  }
  return expectsLowerCase
    ? (val) => !!map[val.toLowerCase()]
    : (val) => !!map[val];
}

export const objectToString = Object.prototype.toString;
export const toTypeString = (value: unknown): string =>
  objectToString.call(value);

export const toRawType = (value: unknown): string => {
  // extract "RawType" from strings like "[object RawType]"
  return toTypeString(value).slice(8, -1);
};

export const isPlainObject = (val: unknown): val is object =>
  toTypeString(val) === "[object Object]";

export const isIntegerKey = (key: unknown) =>
  isString(key) &&
  key !== "NaN" &&
  key[0] !== "-" &&
  "" + parseInt(key, 10) === key;

const cacheStringFunction = <T extends (str: string) => string>(fn: T): T => {
  const cache: Record<string, string> = Object.create(null);
  return ((str: string) => {
    const hit = cache[str];
    return hit || (cache[str] = fn(str));
  }) as T;
};

/**
 * @private
 */
export const capitalize = cacheStringFunction(<T extends string>(str: T) => {
  return (str.charAt(0).toUpperCase() + str.slice(1)) as Capitalize<T>;
});

export const def = (obj: object, key: string | symbol, value: any) => {
  Object.defineProperty(obj, key, {
    configurable: true,
    enumerable: false,
    value,
  });
};
