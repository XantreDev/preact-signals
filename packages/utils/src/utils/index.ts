export { untracked } from "@preact-signals/unified-signals";
export * from "./explicitFalsy";
export * from "./getter";
export * from "./reactive";
export * from "./setter";
export * from "./toSolidApi";
export * from "./type";
export { reaction, rafReaction } from "./reaction";
