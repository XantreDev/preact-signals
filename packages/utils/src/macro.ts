import { ReactiveRef } from "./lib";

console.log(
  "To use macro, you need to use babel plugin. Check out the README for more info."
);
/**
 * This function is compile time shorthand for `$(() => value)`
 *
 * @example
 * ```tsx
 * import { $$ } from "@preact-signals/utils/macro";
 *
 * // This is equivalent to:
 * // const a = $(() => 1)
 * const a = $$(1)
 * ```
 *
 * @example A more complex example
 * ```tsx
 * import { $$ } from "@preact-signals/utils/macro";
 *
 * const a = signal(1)
 * const sig = signal(1)
 * const b = $$(a.value + sig.value)
 *
 * effect(() => {
 *   console.log(b.value)
 * })
 *
 * // logs 3
 * sig.value = 2
 *
 * // logs 4
 * a.value = 2
 * ```
 */
export const $$: <T>(value: T) => ReactiveRef<T> = () => {
  throw new Error("This function is only available at compile time");
};

export const $useState = <T>(value: T): T => {
  throw new Error("This function is only available at compile time");
};
export const $linkedState = <T>(value: T): T => {
  throw new Error("This function is only available at compile time");
};
