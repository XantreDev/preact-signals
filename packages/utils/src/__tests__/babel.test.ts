import { it, describe } from "vitest";
import { format as _format } from "prettier";
import { transform } from "@babel/core";
import preactSignalsUtilsBabel, { BabelMacroPluginOptions } from "../babel";

const format = (code: string) => _format(code, { parser: "acorn" });

type TestCase = {
  type: "success" | "error";
  name: string;
  input: string;
  output: string;
  isCJS: boolean;
  options: BabelMacroPluginOptions | undefined;
};
const TestCase = {
  makeSuccess: (name: string, input: string, output: string): TestCase =>
    TestCase.makeConfigurable(name, input, output, {}),
  makeConfigurable: (
    name: string,
    input: string,
    output: string,
    params: Partial<Omit<TestCase, "input" | "output" | "name">>
  ): TestCase => ({
    type: params.type ?? "success",
    name,
    input,
    output,
    isCJS: params.isCJS ?? false,
    options: params.options ?? {
      enableStateMacros: true,
    },
  }),
  makeError: (name: string, input: string): TestCase =>
    TestCase.makeConfigurable(name, input, "", { type: "error" }),
};

describe.concurrent("@preact-signals/utils/macro", () => {
  const success = [
    TestCase.makeSuccess(
      "ESM import",
      `
      import { $$ } from "@preact-signals/utils/macro";

      const a = $$(1)
    `,
      `
      import { $ as _$ } from "@preact-signals/utils";
      const a = _$(() => 1);
    `
    ),
    TestCase.makeSuccess(
      "Working inside of scopes",
      `
      import { $$ } from "@preact-signals/utils/macro";

      $$(10)
      {
        const a = $$(1)
        
        const $$ = 0
        
        console.log($$)
      }
      `,
      `
      import { $ as _$ } from "@preact-signals/utils";
      _$(() => 10)
      {
        const a = $$(1)
        const $$ = 0;
        console.log($$);
      }
      `
    ),
    TestCase.makeSuccess(
      "Transforms only resolved as macro: unresolved",
      `$$(10)`,
      `$$(10)`
    ),
    TestCase.makeSuccess(
      "Must remove import event if not used",
      `
      import { $$ } from "@preact-signals/utils/macro";
      `,
      ``
    ),
    TestCase.makeSuccess(
      "Transforms only resolved as macro: declared",
      `
      import {$$} from "@preact-signals/utils/macro";
      $$(10)
      {
        const $$ = () => 10;
        $$(10)
      }
      `,
      `
      import { $ as _$ } from "@preact-signals/utils";
      _$(() => 10);
      {
        const $$ = () => 10;
        $$(10);
      }
      `
    ),
    TestCase.makeSuccess(
      "Correctly handles braces in arrow function if using object",
      `
      import { $$ } from "@preact-signals/utils/macro";

      const a = $$({ a: 1 })
    `,
      `
      import { $ as _$ } from "@preact-signals/utils";
      const a = _$(() => ({
        a: 1,
      }));
    `
    ),
    TestCase.makeConfigurable(
      "CJS import",
      `
      const { $$ } = require("@preact-signals/utils/macro");

      const a = $$(1)
      `,
      `
      var _$ = require("@preact-signals/utils").$;
      const a = _$(() => 1);
      `,
      { isCJS: true }
    ),
    TestCase.makeConfigurable(
      "CJS import with multiple imports",
      `
      const { $$, $state } = require("@preact-signals/utils/macro");
      $$(10)
      `,
      `
      var _$ = require("@preact-signals/utils").$;
      const { $state } = require("@preact-signals/utils/macro");
      _$(() => 10)
      `,
      {
        isCJS: true,
        options: {
          enableStateMacros: false,
        },
      }
    ),
    TestCase.makeSuccess(
      "nested macro",
      `
      const {$$} = require("@preact-signals/utils/macro");

      $$($$(1))
      `,
      `
      import { $ as _$ } from "@preact-signals/utils";
      _$(() => _$(() => 1));
      `
    ),
    TestCase.makeSuccess(
      "is not breaking directives",
      `
        'use client';
        'use strict';
        
        import { $$ } from "@preact-signals/utils/macro";

        const a = $$(1)
      `,
      `
        'use client';
        'use strict';
        
        import { $ as _$ } from "@preact-signals/utils";
        const a = _$(() => 1)
        `
    ),
    TestCase.makeSuccess(
      "is not break other imports",
      `
        import React from 'react';
        import {readFileSync} from 'fs';
        import * as path from 'path';
      `,
      `
        import React from 'react';
        import {readFileSync} from 'fs';
        import * as path from 'path';
      `
    ),
    TestCase.makeConfigurable(
      "is not break other imports (CJS)",
      `
        const React = require('react');
        const {readFileSync} = require('fs');
        const path = require('path');
      `,
      `
        const React = require('react');
        const {readFileSync} = require('fs');
        const path = require('path');
      `,
      { isCJS: true }
    ),
    TestCase.makeSuccess(
      "Replaces $state references",
      `
      import { $state } from "@preact-signals/utils/macro";
      const _ = () => {
        let a = $state(0)
        a += 10
        a.value += 10
        a
        a.value
      }
      `,
      `
      const _ = () => {
        let a = 0
        a.value += 10
        a.value.value += 10
        a.value
        a.value.value
      }
      `
    ),
  ];

  for (const { input, isCJS, name, options, output } of success) {
    it(name, async ({ expect }) => {
      expect(
        await format(
          transform(input, {
            plugins: [[preactSignalsUtilsBabel, options]],
            sourceType: isCJS ? "script" : "module",
          })?.code!
        )
      ).toEqual(await format(output));
    });
  }
  const fail = [
    TestCase.makeError(
      "Throws error if not a CallExpression",
      `
      import { $$ } from "@preact-signals/utils/macro";

      const a = $$;
    `
    ),
    TestCase.makeError(
      "Throws error if callExpression called with multiple arguments",
      `
      import { $$ } from "@preact-signals/utils/macro";

      const a = $$(1, 2);
    `
    ),
    TestCase.makeError(
      "Throws error if callExpression called with no arguments",
      `
      import { $$ } from "@preact-signals/utils/macro";

      const a = $$();
    `
    ),
    TestCase.makeError(
      "Throws error if used with spread argument",
      `
      import { $$ } from "@preact-signals/utils/macro";

      const a = $$(...[1]);
    `
    ),
    TestCase.makeError(
      "Throws error if `$state` used with var for variable declaration",
      `
      import { $state } from "@preact-signals/utils/macro";
      const _ = () => {
        var a = $state(0)
      }
      `
    ),
    TestCase.makeError(
      "Throws error if using $state outside of function",
      `
      import { $state } from "@preact-signals/utils/macro";
      let a = $state(0)
      `
    ),
    TestCase.makeConfigurable(
      "CJS cannot rest pattern in require",
      `
      const { $$, ...a } = require("@preact-signals/utils/macro");`,
      ``,
      { isCJS: true }
    ),
    TestCase.makeError(
      "Throws error if $bindedState assigned to a variable",
      `
      import { $bindedState } from "@preact-signals/utils/macro";
      const _ = () => {
        let a = $bindedState(0)
        a += 10
      }
      `
    ),
    // TODO: throw even has no import of known macro
    TestCase.makeError(
      "Throws if imports unknown macro",
      `
      import { $unknown, $$ } from "@preact-signals/utils/macro";
      `
    ),
    // TODO: throw even has no import of known macro
    TestCase.makeError(
      "Throws if imports unknown macro (CJS)",
      `
      const { $unknown, $$ } = require("@preact-signals/utils/macro")
      `
    ),
  ];

  for (const { input, isCJS, name, options } of fail) {
    it(name, async ({ expect }) => {
      expect(() => {
        try {
          transform(input, {
            plugins: [[preactSignalsUtilsBabel, options]],
            sourceType: isCJS ? "script" : "module",
          });
        } catch (e) {
          // console.log(e);
          throw e;
        }
      }).toThrowError();
    });
  }
});
