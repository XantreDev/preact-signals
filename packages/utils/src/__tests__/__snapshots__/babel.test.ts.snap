// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`@preact-signals/utils/macro > $derived transforms correctly 1`] = `
"import { computed as _computed } from "@preact-signals/utils/macro-helper";
const state = _computed(() => 10);
"
`;

exports[`@preact-signals/utils/macro > $useDerived transforms correctly 1`] = `
"import { useComputed as _useComputed } from "@preact-signals/utils/macro-helper";
const _ = () => {
  const state = _useComputed(() => 10);
};
"
`;

exports[`@preact-signals/utils/macro > CJS import 1`] = `
"var _$ = require("@preact-signals/utils").$;
const a = _$(() => 1);
"
`;

exports[`@preact-signals/utils/macro > CJS import with multiple imports 1`] = `
"var _$ = require("@preact-signals/utils").$;
const { $useState } = require("@preact-signals/utils/macro");
_$(() => 10);
"
`;

exports[`@preact-signals/utils/macro > Correctly handles braces in arrow function if using object 1`] = `
"import { $ as _$ } from "@preact-signals/utils";
const a = _$(() => ({
  a: 1,
}));
"
`;

exports[`@preact-signals/utils/macro > Correctly work with references in object shorthand 1`] = `
"import { useDeepSignal as _useDeepSignal } from "@preact-signals/utils/hooks";
const _ = () => {
  let a = _useDeepSignal(() => 0);
  let b = _useDeepSignal(() => 0);
  const c = _useDeepSignal(() => 0);
  return {
    a: a.value,
    b: b.value,
    c: c.value,
  };
};
"
`;

exports[`@preact-signals/utils/macro > ESM import 1`] = `
"import { $ as _$ } from "@preact-signals/utils";
const a = _$(() => 1);
"
`;

exports[`@preact-signals/utils/macro > Must remove import event if not used 1`] = `""`;

exports[`@preact-signals/utils/macro > Replaces $useState references 1`] = `
"import { useDeepSignal as _useDeepSignal } from "@preact-signals/utils/hooks";
import { useSignalOfState as _useSignalOfState } from "@preact-signals/utils/hooks";
const _ = () => {
  let a = _useDeepSignal(() => 0);
  let b = _useDeepSignal(() => 0);
  const c = _useSignalOfState(0);
  a.value += 10;
  a.value.value += 10;
  a.value;
  a.value.value;
  b.value += 10;
  c.value.value += 10;
};
"
`;

exports[`@preact-signals/utils/macro > Top level macro works 1`] = `
"import { deepSignal as _deepSignal } from "@preact-signals/utils";
let a = _deepSignal(0);
effect(() => {
  console.log(a.value);
});
a.value += 10;
"
`;

exports[`@preact-signals/utils/macro > Transforms only resolved as macro: declared 1`] = `
"import { $ as _$ } from "@preact-signals/utils";
_$(() => 10);
{
  const $$ = () => 10;
  $$(10);
}
"
`;

exports[`@preact-signals/utils/macro > Transforms only resolved as macro: unresolved 1`] = `
"$$(10);
"
`;

exports[`@preact-signals/utils/macro > Working inside of scopes 1`] = `
"import { $ as _$ } from "@preact-signals/utils";
_$(() => 10);
{
  const a = $$(1);
  const $$ = 0;
  console.log($$);
}
"
`;

exports[`@preact-signals/utils/macro > is not break other imports (CJS) 1`] = `
"const React = require("react");
const { readFileSync } = require("fs");
const path = require("path");
"
`;

exports[`@preact-signals/utils/macro > is not break other imports 1`] = `
"import React from "react";
import { readFileSync } from "fs";
import * as path from "path";
"
`;

exports[`@preact-signals/utils/macro > is not breaking directives 1`] = `
""use client";
"use strict";

import { $ as _$ } from "@preact-signals/utils";
const a = _$(() => 1);
"
`;

exports[`@preact-signals/utils/macro > nested macro 1`] = `
"import { $ as _$ } from "@preact-signals/utils";
_$(() => _$(() => 1));
"
`;
